include/libdnf/utils/exception.hpp

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright (C) 2020 Red Hat, Inc.
+
+This file is part of libdnf: https://github.com/rpm-software-management/libdnf/
+
+Libdnf is free software: you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation, either version 2 of the License, or
+(at your option) any later version.
+
+Libdnf is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with libdnf.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef LIBDNF_UTILS_EXCEPTION_HPP
+#define LIBDNF_UTILS_EXCEPTION_HPP
+
+
+#include <stdexcept>
+#include <system_error>
+
+namespace libdnf {
+
+/// Base class of libdnf exceptions. Each exception define at least domain_name, name, and description.
+class Exception : public std::runtime_error {
+public:
+    using runtime_error::runtime_error;
+    virtual const char * get_domain_name() const noexcept { return "libdnf"; }
+    virtual const char * get_name() const noexcept { return "Exception"; }
+    virtual const char * get_description() const noexcept { return ""; }
```

Comment:
@lukash I don't ignore constructors taking a custom error description. But I don't reimplement wheel. I used `using runtime_error::runtime_error;` so constructors from `std::runtime_error` (that require user text) are used.
And this description is accessible using what() method as you expect.

The `get_description()` returns static description of the problem (it can be simple translated etc). I suppose that `what()` will be used for dynamic text which is known only in runtime. Eg. filename if file can't be opened.

But you are right, empty message is not ideal. I will add some text like "General libdnf exception".

Reply:
I've actually managed to miss the `using runtime_error::runtime_error;` above, sorry about that...

But anyway, the `what()` is the description, why are you adding another one? That's why I got the impression you're ignoring `what()`... There should be a single message for the error, not multiple ones. How does the user make sense of multiple error descriptions?

And for the `SystemError` below you are actually overriding the `what()` method, so you are basically dropping the built-in dynamic description functionality in favor of your static one.

I see no sense in doing it this way, it seems overly complex to me and makes customizing the error messages hard. I've already set up the base exception classes (as well as some specific exceptions based on them) in dnf 4 in these 3 commits:
https://github.com/rpm-software-management/libdnf/commit/e48f0c12f1f1093de17bea3164791744f09c56d1
https://github.com/rpm-software-management/libdnf/commit/f2aea45236472e75e25b2bdad0857e4625be79a0
https://github.com/rpm-software-management/libdnf/commit/103b834309ecdc1fdc37218e6e8f73c9819b872b

And I'd like to use the same model in dnf 5, unless there's a problem with it (but you reviewed those patches and didn't object...). The patches also showcase how to customize the exception descriptions in the constructor hierarchy (in https://github.com/rpm-software-management/libdnf/commit/103b834309ecdc1fdc37218e6e8f73c9819b872b), where you can add static parts of the description, the `errno` text or whatever else you want.

FWIW my `Error` is roughly equivalent to your `RuntimeError` and my `Exception` is roughly equivalent to your `LogicError` (I prefer my naming :grin: )

> But you are right, empty message is not ideal. I will add some text like "General libdnf exception".

The point here is you don't want to use such exception ever, because it contains no information all. There is always some information about the error you can give.

Reply:
@lukash 
> And for the SystemError below you are actually overriding the what() method, so you are basically dropping the built-in dynamic description functionality in favor of your static one.

This is error. I will fix it.

> And I'd like to use the same model in dnf 5, unless there's a problem with it (but you reviewed those patches and didn't object...).

Yes. Because I didn't want big changes in the exceptions design in DNF4.

> But anyway, the what() is the description, why are you adding another one? That's why I got the impression you're ignoring what()... There should be a single message for the error, not multiple ones. How does the user make sense of multiple error descriptions?

May be, it seems as nonsense to you now. I'm preparing another part of the exceptions design.
The point is that I need to convert catched exception to the human readable type name -> and it can be written to logfile or console, ...
There exists mechanism in C++ for convert type to text description `typeid(type_for_transform_to_name).name()`. But the result of this is not standardized. Some compilers store only name another mangled name, ...
So when the  exception is catched, it can be converted to the human readable text with description and in "what()" will be additional runtime information.

I'm also plan for exception nesting.

Question is if we need to store description `get_description()` method to the exception class. Or the "get_name()" is sufficient and user/program can open external book with description of each exception.

Reply:
I intentionally didn't comment on the name/domain attributes, as I am not sure of their usefulness as of now, but I'm not opposed to them if they are useful for something.

It's the description I don't like, as I said, the `what()` is the description.

Ultimately, you will be composing one line of text with all these data you are introducing. I am composing the same single line of text in the commits I've posted earlier, and note that line of text contains multiple dynamic elements (the filename of the sqlite DB, the error text, the SQL statement (although that's multi-line IIRC :sunglasses: )) and it's working well. Your design doesn't seem to allow this...

Anyway lets have a chat about it sometime :slightly_smiling_face: